airbyte-workers/src/main/java/io/airbyte/workers/internal/EmptyAirbyteSource.java

```diff
@@ -61,4 +121,115 @@ public void cancel() throws Exception {
+    if (hasEmittedState.get()) {
+      return Optional.empty();
+    } else {
+      hasEmittedState.compareAndSet(false, true);
+      return Optional.of(getNullGlobalMessage(streamsToReset, stateWrapper.get().getGlobal()));
+    }
+  }
+
+  private Optional<AirbyteMessage> emitLegacyState() {
+    if (hasEmittedState.get()) {
+      return Optional.empty();
+    } else {
+      hasEmittedState.compareAndSet(false, true);
+      return Optional.of(new AirbyteMessage().withType(Type.STATE)
+          .withState(new AirbyteStateMessage().withStateType(AirbyteStateType.LEGACY).withData(Jsons.emptyObject())));
+    }
+  }
+
+  private boolean isResetAllStreamsInCatalog(final WorkerSourceConfig sourceConfig) {
+    final Set<StreamDescriptor> catalogStreamDescriptors = sourceConfig.getCatalog().getStreams().stream().map(
+        configuredAirbyteStream -> new StreamDescriptor()
+            .withName(configuredAirbyteStream.getStream().getName())
+            .withNamespace(configuredAirbyteStream.getStream().getNamespace()))
+        .collect(Collectors.toSet());
+    final Set<StreamDescriptor> configStreamDescriptors = new HashSet<>(streamsToReset);
+
+    return catalogStreamDescriptors.equals(configStreamDescriptors);
+  }
+
+  private AirbyteMessage getNullStreamStateMessage(final StreamDescriptor streamsToReset) {
+    return new AirbyteMessage()
+        .withType(Type.STATE)
+        .withState(
+            new AirbyteStateMessage()
+                .withStateType(AirbyteStateType.STREAM)
+                .withStream(
+                    new AirbyteStreamState()
+                        .withStreamDescriptor(new io.airbyte.protocol.models.StreamDescriptor()
+                            .withName(streamsToReset.getName())
+                            .withNamespace(streamsToReset.getNamespace()))
+                        .withStreamState(null)));
+  }
+
+  private AirbyteMessage getNullGlobalMessage(final Queue<StreamDescriptor> streamsToReset, final AirbyteStateMessage currentState) {
+    final AirbyteGlobalState globalState = new AirbyteGlobalState();
+    globalState.setStreamStates(new ArrayList<>());
+
+    currentState.getGlobal().getStreamStates().forEach(existingState -> globalState.getStreamStates()
+        .add(
+            new AirbyteStreamState()
+                .withStreamDescriptor(existingState.getStreamDescriptor())
+                .withStreamState(
+                    streamsToReset.contains(new StreamDescriptor()
+                        .withName(existingState.getStreamDescriptor().getName())
+                        .withNamespace(existingState.getStreamDescriptor().getNamespace())) ? null : existingState.getStreamState())));
+
+    // If all the streams in the current state have been reset, we consider this to be a full reset, so
+    // reset the shared state as well
+    if (currentState.getGlobal().getStreamStates().size() == globalState.getStreamStates().stream()
```

Comment:
I thought we always expected all the streams for a Global state. In this case, wouldn't this be always true regardless of whether we are performing a global reset?
Comparing the size of `streamsToReset` to the catalog feels more robust in this case.